staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -29,6 +29,44 @@ type Token interface {
 	AuthenticateToken(ctx context.Context, token string) (*Response, bool, error)
 }
 
+func WrapInInstrumentedRequest(req Request, authenticatorName string) *InstrumentedRequest {
+	return &InstrumentedRequest{Name: authenticatorName, delegate: req}
+}
+
+// InstrumentedRequest wraps Request and allows instrumentation to identify authenticators via the Name field.
+type InstrumentedRequest struct {
+	// Name identifies the delegate authenticator.
+	Name     string
+	delegate Request
+}
+
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *InstrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	if err != nil {
+		err = InstrumentedError{Name: i.Name, delegate: err}
+	}
+
+	if resp != nil {
+		resp.AuthenticatorName = i.Name
```

Comment:
Is it possible that another authenticator has already set the authenticator name on the response?

Reply:
I think it is (ex. union).
Added a check.